snapshots/overlay/overlayutils/check.go

```diff
@@ -198,3 +200,97 @@ func NeedsUserXAttr(d string) (bool, error) {
 	}
 	return true, nil
 }
+
+// SupportsIDMappedMounts tells if this kernel supports idmapped mounts for overlayfs
+// or not.
+//
+// This function returns error whether the kernel supports idmapped mounts
+// for overlayfs or not, i.e. if e.g. -ENOSYS may be returned as well as -EPERM.
+// So, caller should check for (true, err == nil), otherwise treat it as there's
+// no support from the kernel side.
+func SupportsIDMappedMounts() (bool, error) {
+	// Fast path
+	fiveDotNineteen := kernel.KernelVersion{Kernel: 5, Major: 19}
+	if ok, err := kernel.GreaterEqualThan(fiveDotNineteen); err == nil && ok {
+		return true, nil
+	}
+
+	// Do slow path, because idmapped mounts may be backported to older kernels.
+	uidMap := syscall.SysProcIDMap{
+		ContainerID: 0,
+		HostID:      666,
+		Size:        1,
+	}
+	gidMap := syscall.SysProcIDMap{
+		ContainerID: 0,
+		HostID:      666,
+		Size:        1,
+	}
+	td, err := os.MkdirTemp("", "ovl-idmapped-check")
+	if err != nil {
+		return false, fmt.Errorf("failed to create check directory: %w", err)
+	}
+	defer func() {
+		if err := os.RemoveAll(td); err != nil {
+			log.L.WithError(err).Warnf("failed to remove check directory %s", td)
+		}
+	}()
+
+	for _, dir := range []string{"lower", "upper", "work", "merged"} {
+		if err = os.Mkdir(filepath.Join(td, dir), 0755); err != nil {
+			return false, fmt.Errorf("failed to create %s directory: %w", dir, err)
+		}
+	}
+	defer func() {
+		if err = os.RemoveAll(td); err != nil {
+			log.L.WithError(err).Warnf("failed remove overlay check directory %s", td)
+		}
+	}()
+
+	if err = os.Lchown(filepath.Join(td, "upper"), uidMap.HostID, gidMap.HostID); err != nil {
```

Comment:
Hmm, lchown won't follow a symlink, right? Why do we want to change the symlink? 

This will fail with new kernels (from https://lwn.net/SubscriberLink/942954/4a108b4f66074f2a/ ):

>  The kernel will no longer allow the changing of permissions on symbolic links. 

Do we really need this?

Reply:
I'm really not sure that there's a real case for that, but even if there isn't, I think it should be fixed separately, since `main` branch has this too (I didn't change it), so we're risking to break backward compatibility if there's a case for that, but it seems you're right, there might be issues with this later 

I think maintainers can provide more info on this (@fuweid @AkihiroSuda @dmcgowan @kzys)

Reply:
@rata let's keep it as it is. And the upper in this context(function) is unlikely to be symlink right?

Reply:
Yes, let's keep it like that if it is already like that in master